### ISSUE TITLE
Enderpearl bow projectile error

### ISSUE THREAD

**Issue by Brottweiler** [Electroid/PGM#57]
Read: https://github.com/Electroid/PGM/issues/57#issuecomment-551205212

~~Add custom projectiles so maps like Ozone works properly, and so we can have cool snowball bows on Glacial Impact;~~

```xml
<modifybowprojectile>
    <projectile>EnderPearl</projectile>
    <velocityMod>2.5</velocityMod>
</modifybowprojectile>
```
```xml
<modifybowprojectile>
    <projectile>Snowball</projectile>
    <velocityMod>2.0</velocityMod>
</modifybowprojectile>
```

**Comment by Electroid** [comment]
I believe this was deprecated in favor of `<projectiles>`, does the `Projectiles` modules allow all the functionality you need?

**Comment by Brottweiler** [comment]
Probably, but I am unsure how to use it. I tried this

```xml
<projectiles>
    <entity>ender_pearl</entity>
    <velocity>2.5</velocity>
</projectiles>
```
and also replaced `entity` with `id` and also tried using the old `EnderPearl` name.

**Comment by Pablete1234** [comment]
http://docs.oc.tc/modules/projectiles

According to the docs, it should be using the `projectile` element for the entity name

```
<modifybowprojectile>
    <projectile>EnderPearl</projectile>
    <velocityMod>2.5</velocityMod>
</modifybowprojectile>
```

**Comment by Brottweiler** [comment]
@Pablete1234 Didn't @Electroid say that is deprecated? So the docs are "outdated" in this case, for 1.8 PGM?

**Comment by Pablete1234** [comment]
Well, they are as up-to-date as it gets (1.12), if it has been deprecated in this version of pgm, it probably shouldn't have been deprecated, as 1.12 pgm didn't deprecate it (for good reason: no replacement)

**Comment by Brottweiler** [comment]
@Pablete1234 but it was replaced? https://github.com/Electroid/PGM/blob/master/src/main/java/tc/oc/pgm/projectile/ProjectileModule.java

**Comment by Pablete1234** [comment]
https://github.com/Electroid/PGM/blob/master/src/main/java/tc/oc/pgm/modules/ModifyBowProjectileModule.java

**Comment by Brottweiler** [comment]
Ah okay. Well, it works on glacial impact as intended now, but not on Ozone.

```java
[19:28:49 ERROR]: Could not pass event EntityShootBowEvent to PGM v1.8-SNAPSHOT
org.bukkit.event.EventException
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:302) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:74) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:65) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventHandler(SimplePluginManager.java:569) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.dispatchEvent(SimplePluginManager.java:534) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.access$0(SimplePluginManager.java:531) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager$2.callEvent(SimplePluginManager.java:536) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventBody(SimplePluginManager.java:586) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventHandler(SimplePluginManager.java:579) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.dispatchEvent(SimplePluginManager.java:534) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEventSynchronously(SimplePluginManager.java:527) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:521) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:496) [sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.event.CraftEventFactory.callEntityShootBowEvent(CraftEventFactory.java:250) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.ItemBow.a(ItemBow.java:61) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.ItemStack.b(ItemStack.java:527) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.EntityHuman.bU(EntityHuman.java:109) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PlayerConnection.a(PlayerConnection.java:631) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PacketPlayInBlockDig.a(SourceFile:40) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PacketPlayInBlockDig.a(SourceFile:10) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.PlayerConnectionUtils$1.run(SourceFile:13) [sportpaper.jar:git-SportPaper-"374af6d"]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_222]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]
        at net.minecraft.server.v1_8_R3.SystemUtils.a(SourceFile:44) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.processTasks(MinecraftServer.java:1724) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.DedicatedServer.processTasks(DedicatedServer.java:397) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.B(MinecraftServer.java:882) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.DedicatedServer.B(DedicatedServer.java:365) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.A(MinecraftServer.java:821) [sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.MinecraftServer.run(MinecraftServer.java:722) [sportpaper.jar:git-SportPaper-"374af6d"]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
Caused by: java.lang.NullPointerException
        at net.minecraft.server.v1_8_R3.EntityProjectile.<init>(EntityProjectile.java:31) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at net.minecraft.server.v1_8_R3.EntityEnderPearl.<init>(EntityEnderPearl.java:19) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.createEntity(CraftWorld.java:1104) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.spawn(CraftWorld.java:1319) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at org.bukkit.craftbukkit.v1_8_R3.CraftWorld.spawn(CraftWorld.java:1034) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        at tc.oc.pgm.modules.ModifyBowProjectileMatchModule.changeBowProjectile(ModifyBowProjectileMatchModule.java:53) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_222]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_222]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_222]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_222]
        at org.bukkit.plugin.java.JavaPluginLoader$1.execute(JavaPluginLoader.java:300) ~[sportpaper.jar:git-SportPaper-"374af6d"]
        ... 30 more
```

```xml
<modifybowprojectile>
    <projectile>EnderPearl</projectile>
    <velocityMod>2.5</velocityMod>
</modifybowprojectile>
```

**Comment by Pugzy** [comment]
Code at `ModifyBowProjectileMatchModule.java:53`
```java
52        Projectile oldEntity = (Projectile) event.getProjectile();
53        newProjectile = world.spawn(oldEntity.getLocation(), this.cls);
54        event.setProjectile(newProjectile);
```

Maybe related https://hub.spigotmc.org/jira/browse/SPIGOT-2239 and `Resolved in 1.9.3`

**Comment by Pablete1234** [comment]
It's missing a projectile source. The vanilla source is patched to have a `this.projectileSource = (ProjectileSource) entityliving.getBukkitEntity();`, the issue is bukkit/spigot, in createEntity, is using `new EntityEnderPearl(this.world, (EntityLiving)null);`

**Comment by Electroid** [comment]
Is there a fix here that doesn't involve patching Bukkit?

**Comment by Electroid** [comment]
Fixed in #106
